### PR TITLE
Add functionality to construct a PyTorch Model from a pipeline

### PR DESCRIPTION
**Pull request by vladislavalerievich**
This pull request introduces a script for building and evaluating a simple PyTorch model on the MNIST dataset. The script provides functionality for constructing the neural network model from a pipeline, training the model, and evaluating its performance.

> **_NOTE:_** The new functionality is not tied specifically to PyTorch. If we were to replace all instances of torch (e.g., `torch.nn.Linear`) in the pipeline with instances from another framework, the functionality would remain unchanged.

#### Changes Made:
- Added a new script `pytorch_example.py` for building and evaluating a PyTorch model on the MNIST dataset.
- Implemented classes `MatchDimensions` and `MatchChosenDimensions` to handle dimension matching between layers.
- Implemented `build_model_from_pipeline` function for building a PyTorch model from the provided pipeline.
- Added unit tests.
 
#### What does this implement/fix? Explain your changes:
This PR implements a new feature to build and evaluate a PyTorch model on the MNIST dataset. The script allows users to define a pipeline for constructing the model architecture, train the model with specified hyperparameters, and evaluate its accuracy.

#### Minimal Example / How should this PR be tested?
To test this PR, follow these steps:
1. Run the `pytorch_example.py` script.
2. Ensure that the PyTorch model is constructed and trained on the MNIST dataset.
3. Evaluate the model's accuracy on the test dataset.
4. Verify that the model performs satisfactorily according to the defined metrics.

### CODE REVIEW

File: src/amltk/exceptions.py [hunk 1]
```diff
@@ -110,3 +110,50 @@ def __str__(self) -> str:
             f"Duplicate names found in {self.node.name} and can't be handled."
             f"\nnodes: {[n.name for n in self.node.nodes]}."
         )
+
+
+class MatchDimensionsError(Exception):
+    """An exception raised for errors related to matching dimensions in a pipeline."""
+
+    def __init__(self, layer_name: str, param: str | None) -> None:
```
Comment (eddiebergman) [inline src/amltk/exceptions.py:118]: Just as a note, the expected definition of an Exception is `def __init__(*args: Any)`. While we can be more specific here, it sort of breaks a normal contract of implmeneting exceptions. Don't really mind here, just letting you know.

```python
class MatchDimenionError(Exception):
    def __init__(self, layer_name: str, param: str | None, *args: Any) -> None:
    	super.__init__(..., *args)
```

File: src/amltk/pipeline/node.py [hunk 2]
```diff
@@ -468,15 +468,34 @@ def linearized_fidelity(self, value: float) -> dict[str, int | float | Any]:
...
-    def iter(self) -> Iterator[Node]:
-        """Iterate the the nodes, including this node.
+    def iter(self, skip_unchosen: bool = False) -> Iterator[Node]:
+        """Recursively iterate through the nodes starting from this node.
+
+        This method traverses the nodes in a depth-first manner, including
+        the current node and its children nodes.
+
+        Args:
+            skip_unchosen (bool): Flag to skip unchosen nodes in Choice nodes.
```
Comment (eddiebergman) [inline src/amltk/pipeline/node.py:478]: I guess your IDE is putting these in automatically? If not, there's no need to for documentation purposes as the documentation engine will automatically document the type from the function signature.

Don't really need to change it but one reason for changing it is that the documentation and the code doesn't go out of sync, i.e. if the signature is changed from something like `bool | None = False`, then the docstring doesn't go out of sync.

---

Side note, I assume running `ruff` on this will complain. It'll ask for `def iter(self, *, skip_unchosen: bool = False) -> Iterator[Node]`, emphasis on the `*` to indicate keyword only.

Main reason being is that from the callers perspective, it dissallows `node.iter(True)`, where someone else reading has no idea what the `True` is doing. Also prevents issues if another boolean argument is added, in which case the user can't accidentally have the wrong thing,

e.g.. imagine the signature was updated to

```python
# Before update
node.iter(False)

# Code is updated
# This silently breaks the user code above
def iter(self, deep: bool = True, skip_unchosen: bool = False):

```

File: src/amltk/pipeline/node.py [hunk 3]
```diff
@@ -468,15 +468,34 @@ def linearized_fidelity(self, value: float) -> dict[str, int | float | Any]:
...
         yield self
+
+        # Iterate through the child nodes
         for node in self.nodes:
-            yield from node.iter()
+            if skip_unchosen and isinstance(node, Choice):
+                # If the node is a Choice and skipping unchosen nodes is enabled
+                chosen_node = node.chosen()
+                if chosen_node:
+                    yield from chosen_node.iter(skip_unchosen)
```
Comment (eddiebergman) [inline src/amltk/pipeline/node.py:495]: And if there was no `Node` chosen, i.e. `chosen_node == None`? Im not entirely sure what to do here but I imagine it should resort to iterating through all it's children? Open to suggestions why not

File: src/amltk/pytorch/builders.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+        Returns:
+            The value of the matching configuration parameter.
+        """
+        layer = pipeline[self.layer_name]
+        layer_config = layer.config
+        if layer_config is None:
+            raise MatchDimensionsError(self.layer_name)
+        value = layer_config.get(self.param)
+        if value is None:
```
Comment (eddiebergman) [inline src/amltk/pytorch/builders.py:52]: This is a good use case of a `Sentinal Pattern`, i.e. when `None` is a valid value. I don't see many cases for this here but to make configs as flexible as possible, I'm trying to leave `None` as a valid value.

```python
_sentinal = object()
value = layer_config.get(self.param, _sentinal)
if value is _sentinal:
   raise ...
   
# Or without it
if self.param not in lay_config:
   raise ... 
```

File: src/amltk/pytorch/builders.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+        Returns:
+            The value of the matching dimension for a chosen node.
+        """
+        chosen_node_name = chosen_nodes.get(self.choice_name)
+
+        try:
+            return self.choices[chosen_node_name]
+        except KeyError:
+            raise MatchChosenDimensionsError(self.choice_name, chosen_node_name)
```
Comment (eddiebergman) [inline src/amltk/pytorch/builders.py:97]: Usually gives a better traceback with this

```python
        except KeyError as e:
            raise MatchChosenDimensionsError(self.choice_name, chosen_node_name) from e
```

File: src/amltk/exceptions.py [hunk 6]
```diff
@@ -110,3 +110,50 @@ def __str__(self) -> str:
             f"Duplicate names found in {self.node.name} and can't be handled."
             f"\nnodes: {[n.name for n in self.node.nodes]}."
         )
+
+
+class MatchDimensionsError(Exception):
```
Comment (eddiebergman) [inline src/amltk/exceptions.py:115]: I see that they're kind of used as KeyError's mostly, can we inherit from that instead?

File: src/amltk/exceptions.py [hunk 7]
```diff
@@ -110,3 +110,50 @@ def __str__(self) -> str:
...
+                f"Parameter '{param}' not found in the configuration.",
+            )
+        else:
+            super().__init__(
+                f"Error in matching dimensions for layer '{layer_name}'."
+                f" Configuration not found.",
+            )
+
+
+class MatchChosenDimensionsError(Exception):
```
Comment (eddiebergman) [inline src/amltk/exceptions.py:137]: Likewise here regarding `KeyError`

File: src/amltk/pytorch/builders.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        Each pipeline has a unique set of chosen nodes, which we collect separately
+        to handle dimension matching between layers with search spaces.
+
+        Args:
+            pipeline: The pipeline containing the model architecture.
+
+        Returns:
+            The names of the chosen nodes in the pipeline.
+        """
+        chosen_nodes_names = {}  # Class variable to store chosen node names
```
Comment (eddiebergman) [inline src/amltk/pytorch/builders.py:112]: Just a note on the comment going out of date

File: src/amltk/pytorch/builders.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        """
+        chosen_node_name = chosen_nodes.get(self.choice_name)
+
+        try:
+            return self.choices[chosen_node_name]
+        except KeyError:
+            raise MatchChosenDimensionsError(self.choice_name, chosen_node_name)
+
+    @staticmethod
+    def collect_chosen_nodes_names(pipeline: Sequential) -> dict:
```
Comment (eddiebergman) [inline src/amltk/pytorch/builders.py:100]: `dict[str, str]`, sorry for being a stickler on types :/

File: src/amltk/pytorch/builders.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        if isinstance(node, Fixed):
+            model_layers.append(node.item)
+
+        # Check if node is a Component with config parameter
+        elif (
+            isinstance(node.item, type)
+            and issubclass(node.item, nn.Module)
+            and node.config
+        ):
+            layer_config = node.config or {}
```
Comment (eddiebergman) [inline src/amltk/pytorch/builders.py:153]: So there's a silent potential bug here in that your assign `layer_config` to point to `node.config` and then update `layer_config` with new keys, essentially inserting new keys into `node.config`. I doubt this is likely a bug that would come up but it goes against the more _immutable_ nature of the pipelines. 

Main thing would be to just copy the `node.config` which you can easily do with a `Mapping/dict` type by just calling `dict` on it.

```python
        elif (
            isinstance(node.item, type)
            and issubclass(node.item, nn.Module)
            and node.config is not None
        ):
        	layer_config = dict(node.config)  # Does a shallow copy so you can update the keys
```

File: src/amltk/__init__.py [hunk 11]
```diff
@@ -77,4 +82,7 @@
     "Trial",
     "TxtLoader",
     "YAMLLoader",
+    "MatchDimensions",
+    "MatchChosenDimensions",
+    "build_model_from_pipeline",
```
Comment (eddiebergman) [inline src/amltk/__init__.py:87]: So `pytorch` is an optional install, so I'm not sure you can include `build_model_from_pipeline` here. A good way to test that all pytorch code is hidden behind an explicit import is to uninstall pytorch and run one of the other examples.

File: tests/pytorch/test_match_chosen_dimensions.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    Fixed,
+    MatchChosenDimensions,
+    Sequential,
+    build_model_from_pipeline,
+)
+from amltk.exceptions import MatchChosenDimensionsError
+from tests.pytorch.common import create_optimizer
+
+
+class TestMatchChosenDimensions:
```
Comment (eddiebergman) [inline tests/pytorch/test_match_chosen_dimensions.py:18]: Why the class? Nothing against it, just never used it personally

File: tests/pytorch/test_match_chosen_dimensions.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        )
+
+        # Verify that the model is constructed correctly
+        assert isinstance(model, torch.nn.Sequential)
+
+        # Conditional check for the second layer's in_features
+        if model[0].out_features == 20:
+            assert model[1].in_features == 20
+        elif model[0].out_features == 10:
+            assert model[1].in_features == 10
```
Comment (eddiebergman) [inline tests/pytorch/test_match_chosen_dimensions.py:74]: Try to avoid conditionals like these in tests where you can. This relies on explicit expectations of how the setup is, which exist somewhere else and could be accidentally changed and this test would not longer test the feature.

```python
# Remove the if's
assert isinstance(model[0].out_features, int)
assert model[0].out_features == model[1].in_features

# or alternatively, state your assumptions about the setup of the test so that if it ever changes, the test doesn't silently stop doing what it's meant to do
assert model[0].out_features in (10, 20)
... what's there already
```

File: tests/pytorch/test_match_dimensions.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    @pytest.fixture(
+        scope="class",
+        params=[
+            MatchDimensions("non-existing-layer", param="out_features"),
+            MatchDimensions("fc1", param="non-existing-param"),
+            MatchDimensions("fc1", param=None),
+            MatchDimensions(layer_name="", param="out_features"),
+        ],
+    )
```
Comment (eddiebergman) [inline tests/pytorch/test_match_dimensions.py:56]: Is this like a parametrized fixture?

File: tests/pytorch/test_match_dimensions.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+        assert isinstance(model[3], torch.nn.Linear)
+        assert model[3].out_features == 10
+
+    def test_match_dimensions_invalid(self, invalid_pipeline: Node) -> None:
+        """Test for invalid pipeline."""
+        optimizer = create_optimizer(invalid_pipeline)
+        trial = optimizer.ask()
+
+        with pytest.raises((MatchDimensionsError, KeyError)):
```
Comment (eddiebergman) [inline tests/pytorch/test_match_dimensions.py:108]: Should we also check for `KeyError` here, if so, why?

**Review by eddiebergman — CHANGES_REQUESTED** [review]
Mostly good, some questions and some changes needed but otherwise solid :)

File: src/amltk/pipeline/node.py (hunk 2 above)
Comment (vladislavalerievich) [reply src/amltk/pipeline/node.py:478]: I've added unnecessary docstrings exactly because the linter complains.

File: src/amltk/pipeline/node.py (hunk 3 above)
Comment (vladislavalerievich) [reply src/amltk/pipeline/node.py:495]: Do you mean, what will be the consequence if a user didn't use the `Choice` class in the pipeline, but passed the `skip_unchosen=False`?

File: src/amltk/__init__.py (hunk 11 above)
Comment (vladislavalerievich) [reply src/amltk/__init__.py:87]: Should I remove it from the global namespace?

File: tests/pytorch/test_match_chosen_dimensions.py (hunk 12 above)
Comment (vladislavalerievich) [reply tests/pytorch/test_match_chosen_dimensions.py:18]: I thought it is a bit cleaner. Should I change it into functions?

File: tests/pytorch/test_match_chosen_dimensions.py (hunk 13 above)
Comment (vladislavalerievich) [reply tests/pytorch/test_match_chosen_dimensions.py:74]: Noted.

File: tests/pytorch/test_match_dimensions.py (hunk 14 above)
Comment (vladislavalerievich) [reply tests/pytorch/test_match_dimensions.py:56]: Yes, exactly. It runs 4 times with different invalid configuration.

File: tests/pytorch/test_match_dimensions.py (hunk 15 above)
Comment (vladislavalerievich) [reply tests/pytorch/test_match_dimensions.py:108]: The `KeyError` raised inside the `Node` class.

File: src/amltk/pipeline/node.py (hunk 2 above)
Comment (eddiebergman) [reply src/amltk/pipeline/node.py:478]: Interseting, is it PyCharm complaining about no type in the docstring or `ruff`? I've never had `ruff` complain to me about missing docstrings before?

File: src/amltk/__init__.py (hunk 11 above)
Comment (eddiebergman) [reply src/amltk/__init__.py:87]: Aye, I've done the same with scikit-learn as well, such that you shouldn't need scikit-learn installed to use `pytorch` things

File: tests/pytorch/test_match_chosen_dimensions.py (hunk 12 above)
Comment (eddiebergman) [reply tests/pytorch/test_match_chosen_dimensions.py:18]: No no, it's fine, just unfamiliar and was wondering if there was a functional reason.
